### ISSUE TITLE
Challenge 11.4 unclear advantage of implementation

### ISSUE THREAD

**Issue by okomarov** [munificent/craftinginterpreters#1141]
The challenge reads:
> Our resolver calculates which environment the variable is found in, but it’s still looked up by name in that map. A more efficient environment representation would store local variables in an array and look them up by index.
>
> Extend the resolver to associate a unique index for each local variable declared in a scope. When resolving a variable access, look up both the scope the variable is in and its index and store that. In the interpreter, use that to quickly access a variable by its index instead of using a map.

The changes in the interpreter are:

```diff
 public class Interpreter implements Expr.Visitor<Object>, Stmt.Visitor<Void> {
-  final Environment globals = new Environment();
+  final Map<String, Object> globals = new HashMap<>();
-  private Environment environment = globals;
+  private Environment environment;
   private final Map<Expr, Integer> locals = new HashMap<>();
+  private final Map<Expr, Integer> slots = new HashMap<>();
```

So all scoped environments become an array except `globals`. However, every time we need to lookup, define or get a variable, we need to get its location from `slots`. This feels like we swapped a map access from the environment to the locations.

### Question
Am I correct in thinking that while we still have the same number of HashMap accesses, the advantage is that the variables are now stored in a contiguous block of memory since the `Environment` is now an array, and that the efficiency is coming from this change in memory layout rather than not having any HashMap access at all?

In other words, accessing `slots` is faster than accessing `values` (old `Environment.java` code):

```
private final Map<Expr, Integer> slots = new HashMap<>();
```
vs
```
private final Map<String, Object> values = new HashMap<>();
```

**Comment by peppedilillo** [comment]
Dears, I think the approach of OP to this could be wrong. However I can't see any "cheap" solution to solve the challenge satisfyingly enough. I'd like to know if I am missing something.
Personally I have thought of two approaches:

1. Changing `Environment` so that [`Environment.values`](https://github.com/munificent/craftinginterpreters/blob/4a840f70f69c6ddd17cfef4f6964f8e1bcd8c3d4/java/com/craftinginterpreters/lox/Environment.java#L11) is an array. This would change the signature of `Environment.get` from taking a token to taking an integer, removing the lookup there. The appropriate integer would be determined _somehow_ by the resolver. This solution seems overly complicated, and I feel it would require some change in how we deal with global variables, possibly having globals analyzed at compile time too, for assigning them an index. To users which worked a solution to challenge 11.3 into their jlox implementation, having the resolver to deal with globals would likely result in a broken REPL in absence of further adjustments.
2. Changing `Interpreter` transforming `Interpreter.locals` into an array list. This would remove the need for a lookup for environment distance in [`Interpreter.lookUpVariable`](https://github.com/munificent/craftinginterpreters/blob/4a840f70f69c6ddd17cfef4f6964f8e1bcd8c3d4/java/com/craftinginterpreters/lox/Interpreter.java#L483).  It would likely require changing expressions in the AST to store their index in `locals` array. Again, this sounds goofy since it requires messing with the expression objects design. 

Is any of these two the right approach to solving the challenge? Am I missing something altogether? Can we have a hint on how we are supposed to solve challenge 11.4?

Shameless tagging for @munificent . If you get to read this, regardless of this particular problem, thank you for this fantastic book!!

**Comment by kandersen** [comment]
@peppedilillo I believe you are on the right track with both of those solutions. There is nothing sacred about the AST: it's parsed once but run possibly many times. It makes sense to pay the cost of parsing and mapping locals to slots once, and then having that index available at execution for quick look-ups. 

In essence, a local variable just becomes a shorthand for local slots. I have an algorithm in mind that will number them by their order of introduction, but that's by no means the only way.
